### DISCUSSION TITLE
got some really poor performance in a relatively large form

### DISCUSSION THREAD

**Discussion by jackielii** [react-hook-form/react-hook-form#3138] [General]
So the shape of my form is like this:

```
{
  subarray: [{
    name,
    checkbox1,
    checkbox2,
    ...
    checkbox8,
  },
  ... // 46 rows
  ]
}
```
so there are 368 checkboxes displaying.

I've got relative good performance in most interactions except one. To show my interactions, I'll start with a screenshot:

![image](https://user-images.githubusercontent.com/360983/95635248-c9f1e980-0a83-11eb-9ed8-9d3f54100df5.png)

Interactions:

1. click on any checkbox sets a nested value of `subarray[index].checkbox?`
2. calculate column checkbox status: none checked / indeterminate / checked
3. when click on the first row of checkboxes, set all of the checkboxes in this column to checked or not checked. This is the problematic bit

My approach is roughly like below:

1. register all fields manually
2. watch fields individually and add onChange to call setValue

```
const Cell = () => {
      const formMethods = useFormContext()
      const { setValue, control } = formMethods
      const key = `fieldsList[${index}].${field}`
      const value = useWatch({ name: key, control })
      return (
        <Checkbox
          checked={value}
          onChange={(_e, value) => {
            setValue(key, value)
            PerformanceHack(formMethods)()
          }}
        />
      )
}
```

the PerformanceHack there is to set the form dirty, because I can't set dirty on the `key` inside the checkbox onChange, it'll trigger re-render of all the checkboxes. Related code:

```
const PerformanceHackKey = '__internal_performance_hack'
const PerformanceHack = ({ setValue }: PartialRequired<UseFormMethods, 'setValue'>) => () =>
  setValue(PerformanceHackKey, PerformanceHackKey, { shouldDirty: true })
...
// somewhere in the main component: 
  useEffect(() => {
    register(PerformanceHackKey)
  }, [register])
```

This works fairly well, only the components I think should be updated are updated. However the checking / unchecking whole column is extreamly slow, >3 seconds without throttling

profiling react components gives me little information as most components did not render at all:

![image](https://user-images.githubusercontent.com/360983/95635180-9c0ca500-0a83-11eb-9acb-be4489e37319.png)

However, using the profiling tool from Chrome reveals there's a lot of time spent on some functions within react-hook-form: (4x cpu throttle applied this below screen capture)

![image](https://user-images.githubusercontent.com/360983/95635151-8c8d5c00-0a83-11eb-8bad-87c6455b98fd.png)

Am I taking the right approach? or is there any room for optimisation inside the react-hook-form lib?

Thanks & Regards,

**Comment by bluebill1049** [comment]
where is the codesandbox?

**Reply by jackielii** [reply]
https://codesandbox.io/s/crazy-cohen-ttg5s

Tried to simplify as much as possible. Please let me know if you'd like it simplified further.

**Comment by jackielii** [comment]
Hi @bluebill1049 I see you have looked into this: https://github.com/react-hook-form/react-hook-form/commit/3a275ddc45fbb931dd364017da26b419fcb99ddf

I ran a benchmark on master, looks like it's about 1.5 seconds faster. However it's still quite slow. We've eliminated the `stringToPath` call from the top of the call graph. But we're still spending a lot time in function call `transformToNestObject`, `getFieldsValues`. Here is the new profiling screenshot:

![image](https://user-images.githubusercontent.com/360983/95690447-c1c6b500-0c0f-11eb-8785-786c861983a7.png)

**Reply by bluebill1049** [reply]
That PR was not meant to fix your perf issue, that was something I wanted to improve for a while.

here is CSB with a lot more components than your example: https://y0gjq.csb.app/

The question is why `transferNestedObjected` gets invoked? I think you probably missing report the information detail
 above, check the individual checkbox was not the problem here:

> when clicking the checkbox which is already been checked column
<img width="1440" alt="Screen Shot 2020-10-12 at 8 58 34 am" src="https://user-images.githubusercontent.com/10513364/95691205-1d715c80-0c69-11eb-8a99-8deda567e72a.png">

> when clicking on the parent column
<img width="1440" alt="Screen Shot 2020-10-12 at 9 00 37 am" src="https://user-images.githubusercontent.com/10513364/95691239-675a4280-0c69-11eb-99cd-a535d22dc5d3.png">

- when you first check the first checkbox to indicate the checkbox is column is been checked
- when you check the group icon on the checkbox to check the entire column

the problem is on the screen there are about 539 cells, and convert them into a nested object is the bottleneck for the perf. I believe is when you click on the parent checkbox to check all of the child nodes, and first time check one of the nodes to show the parent node.

I will do some more play around with this problem.

**Comment by bluebill1049** [comment]
<img width="569" alt="Screen Shot 2020-10-12 at 9 39 37 am" src="https://user-images.githubusercontent.com/10513364/95691950-da19ec80-0c6e-11eb-9b37-8e5b8d4474da.png">

This is definitely the issue, it's basically doing getValues twice,.

**Comment by bluebill1049** [comment]
<img width="432" alt="Screen Shot 2020-10-12 at 9 43 49 am" src="https://user-images.githubusercontent.com/10513364/95692039-70e6a900-0c6f-11eb-81c3-b90a1ccb3f71.png">

with getValues() * 2 (with 500 plus cells) this is the reason why. i would recommend investigating for a fix there.

**Reply by bluebill1049** [reply]
at my end, i think i can improve the perf with transformNestedObjec, (which i will) but its not the bottle neck of the problem here.

**Reply by jackielii** [reply]
I agree, however, from my end, in order to show if the "CheckAll" checkbox is checked or indeterminate or unchecked, I have to watch "fields"  right? I suppose I can expand all the rows of a specific column and watch those. I'll try this approach.

**Reply by bluebill1049** [reply]
I do get what you try to do, perhaps you want to `useWatch` and only return the result once and cache it, so it doesn't run multiple times. also why calling getValues again? when you already have the watch result. I think if you can reduce invoke watch/useWatch  to single invoke, the perf issue should be improved.

**Reply by jackielii** [reply]
Reason I did `getValues` and then useWatch is commented just above `useWatch`:

```
      // reason why watchValues is not same as values is probably because
      // `watch` happened after `register`, and after register, it didn't
      // re-render, so watch never have a chance to get latest values
      // const watchValues = useWatch<Field.AsObject[]>({
```
So somewhere in my real code I have a `useEffect` to retrieve data from server, and after the data comes back from the server, the values changed, but useWatch didn't get the updated value. 

```
  // values retrieved from server in a custom hook
  useEffect(() => {
    reset(values)
  }, [values, reset])
```

So I tried to use `watch` instead of `useWatch`, `watch` indeed is more performant, however, if I use `watch`, clicking on individual cell updates the all the checkboxes

`useWatch`:

![image](https://user-images.githubusercontent.com/360983/95727566-37fc0380-0c72-11eb-8b81-615b4d878cd6.png)

`watch`:

![image](https://user-images.githubusercontent.com/360983/95727671-64178480-0c72-11eb-85f1-4f35c5ee4918.png)

**Reply by bluebill1049** [reply]
I did a quick test, it's working for me with `useWatch`:

https://codesandbox.io/s/react-hook-form-v6-usewatch-forked-c1xr1?file=/src/index.js

**Reply by bluebill1049** [reply]
also question in your real app, how many do you invoke `watch & getValues()`? is there any way you can cache the result of `watch` or `getValues`()?

**Reply by jackielii** [reply]
The codesandbox code has exactly the same amount of watch & getValues. The only bit I left out is the server fetching bit, which I have a useEffect to reset:

Pseudo code:

```
const item = reduxSelect(state)
useEffect(() => {
  fetchFromServer() // will trigger reduxSelect to be called
},[])
```

just tried watch the list of fields per column and it's now an acceptable performance: about .5 second checkAll a columns:

```
      const watchList = useMemo(() => {
        const result: string[] = []
        for (let i = 0; i < data.length; i++) {
          result.push(`fieldsList[${i}].${column.id}`)
        }
        return result
      }, [column.id, data.length])
      // reason why watchValues is not same as values is probably because
      // `watch` happened after `register`, and after register, it didn't
      // re-render, so watch never have a chance to get latest values
      // const watchValues = useWatch<Field.AsObject[]>({
      useWatch<Field.AsObject[]>({
        name: watchList,
        control: formMethods.control,
        defaultValue: empty,
      })
```

**Reply by bluebill1049** [reply]
> The codesandbox code has exactly the same amount of watch & getValues.

I mean how many times it's been invoked? put some console.log there or add a breakpoint, see how many times its hit there. I think we can reduce that 0.5 seconds even more.

**Reply by jackielii** [reply]
I have updated the codesandbox to exactly the same structure I have in real code. https://codesandbox.io/s/crazy-cohen-ttg5s

Also added the console.log to the useWatch calls, it's called twice for each cell because each cell is rendered twice: once with initial value, then with value from server. same for the "checkall", so 50*9*2 = 900 times. However I'm not sure they're relevant? For each cell, I'm watching it's own value, and for checkAll, I'm watching all rows of it's column. How do I reduce it further? or how do I cache it?

**Reply by bluebill1049** [reply]
<img width="1006" alt="Screen Shot 2020-10-12 at 9 46 53 pm" src="https://user-images.githubusercontent.com/10513364/95738109-732f1e80-0cd4-11eb-863f-241d6199d11d.png">

what about useWatch to watch the entire form? cache the watch result and you can skip those 47times of invoke useWatch.

**Reply by jackielii** [reply]
That's the approach I used in the beginning, the problem with that is every click on any checkbox triggering re-render of the entire table, and that's even slower.

So my early conclusion is react-hook-form is faster than React ;)

**Reply by bluebill1049** [reply]
what about doing this instead?

```tsx
const dataRef = useRef({});
dataRef.current = useWatch({ control }) // watch entire form
// next render if dataRef.current is not {} then just read from the value.

// after click reset 
dataRef.current = {};
```

**Reply by jackielii** [reply]
Hmm, I don't follow. The things is I want react-hook-form to tell me what to re-render, setting the data to a ref doesn't really solve the problem...

e.g. when I click on the "checkall" of a column, I'd like the useWatch to re-render only the individual checkboxes. 

What you suggest there if I understand correctly is the useWatch will trigger, and I pass the values to the table, and then in each checkbox I check if value is changed? This is essentially a re-render right?

**Reply by bluebill1049** [reply]
hmmm you are right...

**Reply by bluebill1049** [reply]
i forgot the part of re-render.

**Comment by Aaronius** [comment]
We're running into something similar. The number of fields in our form is rather large (several hundred fields), though we only render a few fields at a time.

It looks like this:

![image](https://user-images.githubusercontent.com/210820/118069935-458d7f00-b362-11eb-9b55-ad10f8894202.png)

Selecting an item in the tree on the left presents different fields on the right.

We have one `useWatch` usage. Every character press takes ~100 milliseconds to render. It appears that the majority of that time is stuck in `getFieldsValues` as a result of `onChange` being called and the `useWatch`.

![image](https://user-images.githubusercontent.com/210820/118069781-052e0100-b362-11eb-8378-786866c66b30.png)

We're in the process of converting from Formik, which didn't have this particular performance problem.

**Reply by bluebill1049** [reply]
can you share your usage with `useWatch` and `onChange`?

**Reply by bluebill1049** [reply]
also provide a CSB if possible, so I can investigate on it.

**Reply by bluebill1049** [reply]
hey Aaronius, can you try with this beta version, see if that will improve the perf. still would be good to know the usage. `react-hook-form@7.6.1-beta.1`

**Reply by Aaronius** [reply]
Sorry @bluebill1049 , but we ran into too many issues with React-Hook-Form and reverted back to Formik (https://github.com/react-hook-form/react-hook-form/discussions/5159 being the biggest problem).

**Reply by bluebill1049** [reply]
That's all good. That issue was been addressed recently as well, but honestly if formik is working for ur project, don't make the switch.
